Alf config utilities

emailweixu opened the pull request:
I only changed some gin.configurable decorations in the code to alf.configurable
in this change so that the example ppo_babyai.py can run successfully.

TODO: modify grid_search.py, train_play_test.py to support alf config

emailweixu committed:
Alf config utilities

I only change some gin.configurable decorations in the code to alf.configurable
in this change so that the example ppo_babyai.py can run successfully.

TODO: modify grid_search.py, train_play_test.py to support alf config

Haichao-Zhang commented on alf/utils/common.py:
```diff
@@ -429,6 +431,100 @@ def get_gin_file():
+def get_conf_file():
+    """Get the configuration file.
+
+    If ``FLAGS.conf`` is not set, find alf_config.py for configured.gin under
```
```alf_config.py for configured.gin```, ```for```  should be ```or```?

emailweixu replied:
fixed

emailweixu committed:
Fix bug of incorrect handling of default arguments

le-horizon commented on alf/utils/common.py:
```diff
@@ -429,6 +431,105 @@ def get_gin_file():
+            for conf_param in conf_params:
+                pos = conf_param.find('=')
+                if pos == -1:
+                    raise ValueError("conf_params")
```
also include the faulty line in the error?

emailweixu replied:
updated.

le-horizon commented on alf/config_util.py:
```diff
@@ -0,0 +1,479 @@
+    .. code-block::
+
+        alf.config('cool_func', cool_arg2='new_value', cool_arg2='another_value)
+        alf.config('Test.func', b=3)
```
Test.func definition is missing?

emailweixu committed:
Fix policy_trainer_test

emailweixu committed:
Update comments

emailweixu committed:
Update comments

hnyu commented on alf/bin/train.py:
```diff
@@ -47,6 +47,7 @@
 from absl import logging
 import gin
 import os
+import runpy
```
not used

emailweixu replied:
removed

hnyu commented on alf/config_helpers.py:
```diff
@@ -0,0 +1,124 @@
+
+The main motivation is to give the access of observation_spec and action_spec,
+which are necessary for config some models. observation_spec and action_spec are
+only available after the environment is craeted. So we create an environment
```
created

emailweixu replied:
Feel it should not be past tense.

hnyu replied:
I mean the typo craeted -> created

hnyu commented on alf/config_helpers.py:
```diff
@@ -0,0 +1,124 @@
+        env = get_env()
+        data_transformer = create_data_transformer(
+            config.data_transformer_ctor, env.observation_spec())
+        observation_spec = data_transformer.transformed_observation_spec
```
observation_spec is not used?

emailweixu replied:
merged with the next line

hnyu commented on alf/config_util.py:
```diff
@@ -0,0 +1,480 @@
+def _add_to_conf_tree(module_path, func_name, para_name, node):
+    """
+    Args:
+        module_path (list[str]):
```
missing description for the arguments.

emailweixu replied:
added

hnyu commented on alf/config_util.py:
```diff
@@ -0,0 +1,480 @@
+    names = []
+    for name in reversed(path[1:]):
+        if not isinstance(tree, dict):
+            raise ValueError("'%s' conflicts with existing config name '%s' "
```
This error message is confusing. Can be `"%s path is not valid; only found a partial path up to %s" % ('.'join(path), '.'.join(reversed(names)))`

emailweixu replied:
I feel the current one seems more appropriate. For example, suppose we want to add "algorithm.vae.test.b", but "vae.test.b" has already been added. The message will be:

'algorithm.vae.test.b' conflicts with existing config name 'vae.test.b'

which seems to a correct description of the problem.

hnyu replied:
I'm actually not sure about the intended usage here. Suppose we have a function called "B.a", and another function "A.B.a". So basically if I add "A.B.a" first and then "B.a", there will be no error. But I can't do in the reversed order?

hnyu commented on alf/config_util.py:
```diff
@@ -0,0 +1,480 @@
+    The decorator can be supplied with parameters to specify the configurable name
+    or supply a whitelist/blacklist:
+
+        @alf.configurable('my_func', whitelist='param2')
```
whitelist=['param2']

emailweixu replied:
fixed

hnyu commented on alf/config_util.py:
```diff
@@ -0,0 +1,480 @@
+    In this case, the name of the configurable is 'MyClass', and both `param1`
+    and `param2` are configurable.
+
+    Note: currently, to maitain the compatibility with gin-config, all the
```
maintain

emailweixu replied:
fixed

hnyu commented on alf/config_util.py:
```diff
@@ -0,0 +1,480 @@
+    and `param2` are configurable.
+
+    Note: currently, to maitain the compatibility with gin-config, all the
+        functions decorated using alf.configurable are auomatically configurable
```
automatically

emailweixu replied:
fixed

hnyu commented on alf/examples/ppo_babyai.py:
```diff
@@ -0,0 +1,140 @@
+
+encoding_dim = 128
+activation = torch.relu_
+fc_layers_params = (256, 256)
```
Will these global variables won't interfere with variables of the same names in python code?

le-horizon commented on alf/config_helpers.py:
```diff
@@ -0,0 +1,124 @@
+    """Get the training environment.
+
+    Note: you need to finish all the config for environments and
+    TrainerConfig.random_seed before using this function.
```
Is there a simple way to ensure this in code?  e.g. assert TrainerConfig is properly set?  Or is the assert in _Config.get_effective_value() already doing that?

emailweixu replied:
If a value is used but configured later, there will be an exception in config1()
```python
raise ValueError(
            "Config '%s' has already been used. You should config "
            "its value before using it." % config_name)
```

le-horizon commented on alf/config_util.py:
```diff
@@ -0,0 +1,480 @@
+
+
+class _Config(object):
+    """Object prepresenting one configurable value."""
```
->representing

emailweixu replied:
fixed

Haichao-Zhang commented on alf/config_helpers.py:
```diff
@@ -0,0 +1,124 @@
+
+
+def get_env():
+    """Get the training environment.
```
global environment?

emailweixu replied:
added.

Haichao-Zhang commented on alf/config_util.py:
```diff
@@ -0,0 +1,480 @@
+
+    Example:
+
+    Asumme we have the following decorated functions and classes:
```
Asumme  -> Assume

emailweixu replied:
fixed.

Haichao-Zhang commented on alf/config_util.py:
```diff
@@ -0,0 +1,480 @@
+
+    .. code-block::
+
+        alf.config('cool_func', cool_arg2='new_value', cool_arg2='another_value)
```
'another_value -> 'another_value'

emailweixu replied:
fixed

Haichao-Zhang commented on alf/config_util.py:
```diff
@@ -0,0 +1,480 @@
+    .. code-block::
+
+        alf.config('cool_func', cool_arg2='new_value', cool_arg2='another_value)
+        alf.config('Test.func', b=3)
```
``Test.func``  missing in the code-block?

emailweixu replied:
changed to Worker.func

Haichao-Zhang commented on alf/config_util.py:
```diff
@@ -0,0 +1,480 @@
+
+        alf.config('cool_func', cool_arg2='new_value', cool_arg2='another_value)
+        alf.config('Test.func', b=3)
+        alf.config('func', b=3)     # 'Test.func' can be uniquely identified by 'func'
```
Test.func

emailweixu replied:
changed to Worker.func

Haichao-Zhang commented on alf/config_util.py:
```diff
@@ -0,0 +1,480 @@
+
+    leaves = _get_all_leaves(tree)
+    if len(leaves) > 1:
+        # only show atmost 3 ambiguous choices
```
atmost -> at most

Haichao-Zhang commented on alf/config_util_test.py:
```diff
@@ -0,0 +1,148 @@
+        # Test duplicated config
+        with self.assertLogs() as ctx:
+            alf.config1('test_func2.c', 15)
+            alf.config1('test_func2.c', 16)
```
This seems to be different from the behavior of gin, where the later config will overwrite the prior one?
By using replacing_existing_config flag, it should be safer.

emailweixu replied:
Right. The reason for this is to be able to correctly implement '--conf_param' commandline flag. In order for --conf_param flag to work, we have to set it before parsing conf file because the conf file may depends on the value configured through conf_param. Hence we cannot allow replacing_existing_config in the conf file.

And it seems to me it is hard to say replacing_existing_config=True is safer than False. Because there is no way to tell which config the user really want. Anyway, a warning will be generated.

le-horizon replied:
Can command line --conf_param override existing configs in the conf file?
My personal habit is to have the conf file being self sufficient (so it can be used standalone), and have conf_param provide ways of customizing configs based on the conf file (e.g. running one off experiments tweaking some configs).
Does that make sense?

emailweixu replied:
Right. --conf_param override existing configs from the conf file.

le-horizon replied:
Then I don't quite understand why some confs in the conf file may depend on --conf_param values?

emailweixu committed:
Address review comments

le-horizon commented on alf/config_util_test.py:
```diff
@@ -0,0 +1,151 @@
+        logging.info("get_operative_configs(): \n%s" % pprint.pformat(
+            alf.get_operative_configs()))
+        logging.info("get_inoperative_configs(): \n%s" % pprint.pformat(
+            alf.get_inoperative_configs()))
```
add asserts for the set of operative config names and inoperative config names equal the correct lists of names.

emailweixu replied:
added.

le-horizon reviewed (commented):
Thanks for the nice and clean change and awesome improvement to configs!
I like the clean format of the new config a lot.

Some nitpicks inlined.

Also, I guess the following aren't supported yet?
 config scope,
 include another config file.

Thanks,
Le

emailweixu committed:
Better handling of name conflicts

Also make the behaviour of random_seed same as before so that tests won't fail.

emailweixu commented:
> Also, I guess the following aren't supported yet?
> config scope,

no config scope. Because we are using python now, it's quite easy to achieve the same effects.

> include another config file.

you can always import another python module

le-horizon commented:
> > Also, I guess the following aren't supported yet?
> > config scope,
> 
> no config scope. Because we are using python now, it's quite easy to achieve the same effects.

Oh, I see, just provide different args to the objects, e.g.
```
a = Class(arg=a); b = Class(arg=b)
```
> 
> > include another config file.
> 
> you can always import another python module

Make sense.

emailweixu committed:
Address more comments

le-horizon reviewed (dismissed):
Haven't closely examined the config_util.py.  The rest LGTM.

Le

emailweixu committed:
Update comments

hnyu commented on alf/config_helpers.py:
```diff
@@ -106,6 +107,8 @@ def get_env():
     if _env is None:
         trainer_config = TrainerConfig(root_dir='')
+        trainer_config.random_seed = set_random_seed(
+            trainer_config.random_seed)
```
If the random seed is set here, we need to guarantee that no other env is created before the first call of get_env(). This might be problematic. I think a better way is still call set_random_seed at the entry of the main function.

emailweixu replied:
moved to entry

hnyu commented on alf/config_util.py:
```diff
@@ -448,6 +471,15 @@ def __init__(self, param1, param2='a default value'):
+    or "uvw.func.a" to refer these two configurable values. You cannot use
+    "func.a" because of the ambiguity. Because of this, you cannot have a config
+    name which is the strict suffix of another config name. For example,
+    "A.Test.arg" and "Test.arg" cannot both be defined.
```
What should we do if there is a global function called "a" we want to configure, but there is also another class function "A.a"?

hnyu replied:
I think it would be good to add a section in ALF notes explaining ALF config name scheme on RTD once this PR is merged.

emailweixu replied:
You can supply a different name:
```python
@configurable("global_a")
def a(arg):
   ...
```
Now the config name is "global_a.arg". This usage is described in the docstring.

emailweixu committed:
Move set_random_seed to main and more comment about name conflict

Haichao-Zhang commented on alf/config_util.py:
```diff
@@ -478,7 +478,21 @@ def __init__(self, param1, param2='a default value'):
+    "A.Test.arg" and "Test.arg" cannot both be defined. You can supply a different
+    name for the function to avoid conflict:
+
+    .. code-block:: pytyon
```
pytyon -> python

Haichao-Zhang commented on alf/config_util.py:
```diff
@@ -478,7 +478,21 @@ def __init__(self, param1, param2='a default value'):
+
+    or
+
+    .. code-block:: pytyon
```
pytyon -> python

Haichao-Zhang commented on alf/config_util.py:
```diff
@@ -478,7 +478,21 @@ def __init__(self, param1, param2='a default value'):
+    name for the function to avoid conflict:
+
+    .. code-block:: pytyon
+        ``@alf.configurable("NewTest")
```
`` can be removed?

Haichao-Zhang commented on alf/config_util.py:
```diff
@@ -478,7 +478,21 @@ def __init__(self, param1, param2='a default value'):
+    or
+
+    .. code-block:: pytyon
+        ``@alf.configurable("B.Test")
```
``

Haichao-Zhang reviewed (commented):
Some minor issues.

emailweixu committed:
Fix docstring
